tpi_rs/src/middleware/serial.rs

```diff
@@ -0,0 +1,253 @@
+//! Handlers for UART connections to/from nodes
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use bytes::BytesMut;
+use futures::stream::{SplitSink, SplitStream};
+use futures::{SinkExt, StreamExt};
+use tokio::sync::Mutex;
+use tokio::task::JoinHandle;
+use tokio_serial::{DataBits, Parity, SerialPortBuilderExt, SerialStream, StopBits};
+use tokio_util::codec::{BytesCodec, Decoder, Framed};
+
+use super::NodeId;
+
+const OUTPUT_BUF_SIZE: usize = 16 * 1024;
+
+type Sink = SplitSink<Framed<SerialStream, BytesCodec>, BytesMut>;
+type Stream = SplitStream<Framed<SerialStream, BytesCodec>>;
+
+#[derive(Debug)]
+pub struct SerialConnections {
+    handlers: Vec<Handler>,
+}
+
+#[derive(Debug)]
+struct Handler {
+    node: usize,
+    sink: Sink,
+    stream: Arc<Mutex<Stream>>,
+    buf: Arc<Mutex<RingBuffer<OUTPUT_BUF_SIZE>>>,
+    worker: Option<JoinHandle<()>>,
+}
+
+#[derive(Debug)]
+struct RingBuffer<const C: usize> {
+    buf: Vec<u8>,
+    idx: usize,
+    len: usize,
+}
+
+impl SerialConnections {
+    pub fn new() -> Result<Self> {
+        let paths = ["/dev/ttyS2", "/dev/ttyS1", "/dev/ttyS4", "/dev/ttyS5"];
+
+        let mut handlers = vec![];
+
+        for (i, path) in paths.iter().enumerate() {
+            handlers.push(Handler::new(i + 1, path)?);
+        }
+
+        Ok(SerialConnections { handlers })
+    }
+
+    pub fn run(&mut self) {
+        for h in &mut self.handlers {
+            h.start_reader();
+        }
+    }
+
+    pub async fn read(&self, node: NodeId) -> String {
+        let idx = node as usize;
+        let handler = &self.handlers[idx];
+        let buf = handler.buf.lock().await.read();
+
+        String::from_utf8_lossy(&buf).to_string()
```

Comment:
what about `utf16` and `Unicode` encoding? Perhaps we should be able to specify in the API an option to use a different encoding and default to UTF8?

Reply:
Actually we should just send bytes as-is, without UTF-8 conversion

Reply:
Actually we can't, and you're right. But it does need to converted to UTF-8 per JSON's [RFC 8259](https://www.rfc-editor.org/rfc/rfc8259) sect. 8.1.